File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/sumNumbers_129.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class sumNumbers_129 {
    int res = 0;

    public int sumNumbers(TreeNode root) {
        helper(root, 0);
        return res;
    }

    private void helper(TreeNode root, int sum) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            res += sum * 10 + root.val;
            return;
        }
        helper(root.left, sum * 10 + root.val);
        helper(root.right, sum * 10 + root.val);
    }
}

File: Clion/ZeroTrac/Tree/sumNumbers_129.cpp
//
// Created by TzashiNorpu on 10/25/2022.
//
#include "tree.h"

using namespace Tree;

int Solution::sumNumbers(TreeNode *root) {
  int sum = 0, res = 0;
  function<void(TreeNode *)> dfs = [&](TreeNode *node) -> void {
      if (!node) return;
      if (!node->left && !node->right) {
        res += (sum * 10 + node->val);
        return;
      }
      sum = sum * 10 + node->val;
      dfs(node->left);
      dfs(node->right);
      sum /= 10;
  };
  return res;
}